Fix build issues on mingw

alexcrichton opened the pull request:
At least in the mingw I'm using, there's no `cc`, but rather `gcc`, so default to `gcc` to avoid having to type `CC=gcc` all the time. Also append to CFLAGS instead of overriding them just in case some extra ones are supplied from the command line.

Finally, the real build issue was `DEVICE_TYPE` not being defined early enough, so define it a bit earlier to get it to build.

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit alexcrichton/libuv@0817121b7f46d13070166084de13c89740f0a339 has the following error(s):
- Commit message must indicate the subsystem this commit changes

The following commiters were not found in the CLA:
- Alex Crichton

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

bnoordhuis commented:
Can you split this into two commits, one for the Makefile, the other for the header issue?  The usual prefix is 'windows:' and please be detailed in the commit log (see 'git log' for inspiration): explain what changed and why.  Don't worry about being wordy, I rather have commit logs with too much detail than too little.  Thanks.

bnoordhuis commented:
> Don't worry about being wordy, I rather have commit logs with too much detail than too little.

And to expand on that, the commit log should be so descriptive that I don't need to look at the patch anymore to find out what changed.

alexcrichton committed:
windows: Tweak Makefile.mingw for easier usage

Default CC to `gcc` instead of `cc` because at least on the mingw installation I
had there was no `cc` exectuable. This avoids having to type `CC=gcc` in all the
`make` commands.

Also append to `CFLAGS` instead of defining if not previously defined. These
flags are required to build libuv, so they should not be overridden if other
extra CFLAGS are supplied via the command line as well.

alexcrichton committed:
windows: Fix compiling on mingw

On older mingw installations, DEVICE_TYPE isn't defined. This was redefined in
windows.h, but it was done after it was already attempted to be used. This moves
the #define for DEVICE_TYPE above the usage of it in windows.h to prevent
compilation failures.

bnoordhuis commented:
Thanks Alex, landed in a9b3fe2 and b647c27.
